framework/include/utils/InputParameterWarehouse.h

```diff
@@ -23,7 +23,32 @@
 class InputParameters;
 class MooseApp;
 
-typedef std::map<std::string, MooseSharedPointer<InputParameters> >::iterator InputParameterIterator;
+/**
+ * A storage container for the name of parameters to be controlled
+ */
+struct InputParametersContainer
+{
+  /// The "system" name (e.g., Kernels)
+  std::string system;
+
+  /// The user-defined tag assigned to the object
+  std::string tag;
+
+  /// The "object" name (generally a sub block in an input file)
+  std::string object;
+
+  /// The name of the parameter to be controlled
+  std::string param;
+
+  /// Input file syntax provided by creating Action
+  std::string syntax;
```

Comment:
Same with `syntax`.  It shouldn't be a separate thing... it should just be one of the tags in the `std::set`